charts/redpanda/chart_template_test.go

```diff
@@ -0,0 +1,272 @@
+			ValuesFile: "./ci/" + name,
+			Assert: func(t *testing.T, b []byte) {
+				testutil.AssertGolden(t, testutil.YAML, path.Join("testdata", "ci", name+".golden"), b)
+			},
+		}
+	}
+	return cases
+}
+
+func VersionGoldenTestsCases(t *testing.T) []TemplateTestCase {
+	// A collection of versions that should trigger all the gates guarded by
+	// "redpanda-atleast-*" helpers.
+	versions := []redpanda.PartialImage{
+		{Tag: ptr.To(redpanda.ImageTag("v22.2.0"))},
+		{Tag: ptr.To(redpanda.ImageTag("v22.3.0"))},
+		{Tag: ptr.To(redpanda.ImageTag("v22.3.14"))},
+		{Tag: ptr.To(redpanda.ImageTag("v22.4.0"))},
+		{Tag: ptr.To(redpanda.ImageTag("v23.1.1"))},
+		{Tag: ptr.To(redpanda.ImageTag("v23.1.2"))},
+		{Tag: ptr.To(redpanda.ImageTag("v23.1.3"))},
+		{Tag: ptr.To(redpanda.ImageTag("v23.2.1"))},
+		{Tag: ptr.To(redpanda.ImageTag("v23.3.0"))},
+		{Tag: ptr.To(redpanda.ImageTag("v24.1.0"))},
+		{Repository: ptr.To("somecustomrepo"), Tag: ptr.To(redpanda.ImageTag("v24.1.0"))},
+	}
+
+	// A collection of features that are protected by the various above version
+	// gates.
+	permutations := []redpanda.PartialValues{
+		{
+			Config: &redpanda.PartialConfig{
+				Tunable: &redpanda.PartialTunableConfig{
+					"log_segment_size_min":  100,
+					"log_segment_size_max":  99999,
+					"kafka_batch_max_bytes": 7777,
+				},
+			},
+		},
+		{
+			Enterprise: &redpanda.PartialEnterprise{License: ptr.To("ATOTALLYVALIDLICENSE")},
+		},
+		{
+			RackAwareness: &redpanda.PartialRackAwareness{
+				Enabled:        ptr.To(true),
+				NodeAnnotation: ptr.To("topology-label"),
+			},
+		},
+	}
+
+	var cases []TemplateTestCase
+	for _, version := range versions {
+		for i, perm := range permutations {
+			values, err := valuesutil.UnmarshalInto[redpanda.PartialValues](perm)
+			require.NoError(t, err)
+
+			values.Image = &version
+
+			name := fmt.Sprintf("%s-%s-%d", ptr.Deref(version.Repository, "default"), *version.Tag, i)
+
+			cases = append(cases, TemplateTestCase{
+				Name:   name,
+				Values: values,
+				Assert: func(t *testing.T, b []byte) {
+					testutil.AssertGolden(t, testutil.YAML, path.Join("testdata", "versions", name+".yaml.golden"), b)
+				},
+			})
+		}
+	}
+	return cases
+}
+
+func DisableCertmanagerIntegration(t *testing.T) []TemplateTestCase {
+	assertNoCerts := func(t *testing.T, b []byte) {
+		// Assert that no Certificate objects are in the resultant
+		// objects when SecretRef is specified AND RequireClientAuth is
+		// false.
+		objs, err := kube.DecodeYAML(b, redpanda.Scheme)
+		require.NoError(t, err)
+
+		for _, obj := range objs {
+			_, ok := obj.(*certmanagerv1.Certificate)
+			// The -root-certificate is always created right now, ignore that
+			// one.
+			if ok && strings.HasSuffix(obj.GetName(), "-root-certificate") {
+				continue
+			}
+			require.Falsef(t, ok, "Found unexpected Certificate %q", obj.GetName())
+		}
+
+		require.NotContains(t, b, []byte(certmanagerv1.CertificateKind))
+	}
+
+	return []TemplateTestCase{
```

Comment:
What do you feel isn't being reflected? Your case should be represented by `disable-cert-manager-fully-specified` which is explicitly specifying a certificate for all listeners and not utilizing the `default` or `external` certificates. You'll need to add in `enabled: false` to your case to make it work as expected/desired.

Reply:
> enabled: false

was the missing link

Reply:
It is a bit counter intuitive, though. Wouldn't that disable TLS altogether? since it seems to be a global toggle.

Reply:
You are talking about taking this: 

```yaml
 tls:
      certs:
        letsencrypt:
          caEnabled: false
          secretRef:
            name: letsencrypt-cert
        selfsigned:
          caEnabled: true
          secretRef:
            name: selfsigned-cert
      enabled: true
```
to this: 

```yaml
 tls:
      certs:
        letsencrypt:
          caEnabled: false
          secretRef:
            name: letsencrypt-cert
        selfsigned:
          caEnabled: true
          secretRef:
            name: selfsigned-cert
      enabled: false
```
right?

Reply:
YAML makes this difficult to communicate well but:

```yaml
tls:
  certs:
    default:
      enabled: false # <--- Disables the "default" certificate which doesn't have a secretRef. 
```

`default` is always present in the TLS map because it's included in the chart's values.yaml. The chart fairly frequently uses `enabled: bool` for toggling features on and off _but_ there's so many occurrences that it's difficult to tell how it actually works.

The typing starts to get a bit weird but we could instead add support for this:
```yaml
tls:
  certs:
    default: null # This would "clear" the default cert more explicitly instead of providing a .enabled check
```